src/persist-types/src/stats2.rs

```diff
@@ -0,0 +1,311 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Aggregate statistics about data stored in persist.
+
+use std::fmt::Debug;
+
+use arrow::array::{
+    Array, BinaryArray, BooleanArray, Float32Array, Float64Array, Int16Array, Int32Array,
+    Int64Array, Int8Array, StringArray, UInt16Array, UInt32Array, UInt64Array, UInt8Array,
+};
+
+use crate::stats::primitive::{
+    truncate_bytes, truncate_string, PrimitiveStats, TruncateBound, TRUNCATE_LEN,
+};
+use crate::stats::{
+    ColumnNullStats, ColumnStatKinds, ColumnarStats, DynStats, NoneStats, OptionStats,
+};
+
+/// A type that can incrementally collect stats from a sequence of values.
+pub trait ColumnarStatsBuilder<T>: Debug {
+    /// Type of [`arrow`] column these statistics can be derived from.
+    type ArrowColumn: arrow::array::Array + 'static;
+    /// The type of statistics the collector finalizes into.
+    type FinishedStats: DynStats;
+
+    /// Create a new instance of `Self` that can be used to collect statistics.
+    fn new() -> Self
+    where
+        Self: Sized;
+
+    /// Derive statistics from a column of data.
+    fn from_column(col: &Self::ArrowColumn) -> Self
+    where
+        Self: Sized;
+    /// Derive statistics from an opaque column of data.
+    ///
+    /// Returns `None` if the opaque column is not the same type as
+    /// [`Self::ArrowColumn`].
+    fn from_column_dyn(col: &dyn arrow::array::Array) -> Option<Self>
+    where
+        Self: Sized,
+    {
+        let col = col.as_any().downcast_ref::<Self::ArrowColumn>()?;
+        Some(Self::from_column(col))
+    }
+
+    /// Include the provided value in the current statistics.
+    fn include(&mut self, val: T);
+
+    /// Finish this collector returning the final aggregated statistics.
+    fn finish(self) -> Self::FinishedStats
+    where
+        Self::FinishedStats: Sized;
+}
+
+/// We collect stats for all primitive types in exactly the same way. This
+/// macro de-duplicates some of that logic.
+///
+/// Note: If at any point someone finds this macro too complext, they should
```

Comment:
```suggestion
/// Note: If at any point someone finds this macro too complex, they should
```

(I don't find it that complex, but I think it's a bit of a smell that we're generating all these `include` impls etc. that are never called...)

Reply:
Totally fair, at the moment I wanted to maintain parity with the other trait impls, but can totally revisit this